File: Day-22/event-upcasting-sample/java/EventUpcastingLedger.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class EventUpcastingLedger {
    public static final class RawEvent {
        private final int sequence;
        private final String accountId;
        private final String eventType;
        private final int schemaVersion;
        private final Map<String, Object> payload;

        public RawEvent(
            int sequence,
            String accountId,
            String eventType,
            int schemaVersion,
            Map<String, Object> payload
        ) {
            this.sequence = sequence;
            this.accountId = accountId;
            this.eventType = eventType;
            this.schemaVersion = schemaVersion;
            this.payload = payload;
        }

        public int getSequence() {
            return sequence;
        }

        public String getAccountId() {
            return accountId;
        }

        public String getEventType() {
            return eventType;
        }

        public int getSchemaVersion() {
            return schemaVersion;
        }

        public Map<String, Object> getPayload() {
            return payload;
        }
    }

    public static final class CanonicalEvent {
        private final int sequence;
        private final String accountId;
        private final String eventType;
        private final int schemaVersion;
        private final int amountCents;
        private final String currency;

        public CanonicalEvent(
            int sequence,
            String accountId,
            String eventType,
            int schemaVersion,
            int amountCents,
            String currency
        ) {
            this.sequence = sequence;
            this.accountId = accountId;
            this.eventType = eventType;
            this.schemaVersion = schemaVersion;
            this.amountCents = amountCents;
            this.currency = currency;
        }

        public int getSequence() {
            return sequence;
        }

        public String getAccountId() {
            return accountId;
        }

        public String getEventType() {
            return eventType;
        }

        public int getSchemaVersion() {
            return schemaVersion;
        }

        public int getAmountCents() {
            return amountCents;
        }

        public String getCurrency() {
            return currency;
        }
    }

    public static final class EventStore {
        private final List<RawEvent> events = new ArrayList<>();
        private int nextSequence = 1;

        public RawEvent append(
            String accountId,
            String eventType,
            int schemaVersion,
            Map<String, Object> payload
        ) {
            if (accountId == null || accountId.isBlank()) {
                throw new IllegalArgumentException("accountId must be non-empty");
            }
            if (eventType == null || eventType.isBlank()) {
                throw new IllegalArgumentException("eventType must be non-empty");
            }
            if (schemaVersion <= 0) {
                throw new IllegalArgumentException("schemaVersion must be positive");
            }

            RawEvent event = new RawEvent(
                nextSequence,
                accountId,
                eventType,
                schemaVersion,
                new LinkedHashMap<>(payload)
            );
            events.add(event);
            nextSequence += 1;
            return event;
        }

        public List<RawEvent> eventsForAccount(String accountId) {
            List<RawEvent> result = new ArrayList<>();
            for (RawEvent event : events) {
                if (accountId.equals(event.getAccountId())) {
                    result.add(event);
                }
            }
            return result;
        }
    }

    public static final class EventUpcaster {
        private final Map<String, Integer> latestVersionByType = new LinkedHashMap<>();

        public EventUpcaster() {
            latestVersionByType.put("FundsDeposited", 2);
            latestVersionByType.put("FundsWithdrawn", 2);
        }

        public CanonicalEvent upcast(RawEvent event) {
            Integer latest = latestVersionByType.get(event.getEventType());
            if (latest == null) {
                throw new IllegalArgumentException("unsupported eventType: " + event.getEventType());
            }
            if (event.getSchemaVersion() > latest) {
                throw new IllegalArgumentException(
                    "unsupported future schema for "
                        + event.getEventType()
                        + ": "
                        + event.getSchemaVersion()
                        + " > "
                        + latest
                );
            }

            int currentVersion = event.getSchemaVersion();
            Map<String, Object> payload = new LinkedHashMap<>(event.getPayload());

            while (currentVersion < latest) {
                payload = upcastOnce(event.getEventType(), currentVersion, payload);
                currentVersion += 1;
            }

            int amountCents = ((Number) payload.get("amount_cents")).intValue();
            String currency = payload.get("currency").toString();

            return new CanonicalEvent(
                event.getSequence(),
                event.getAccountId(),
                event.getEventType(),
                currentVersion,
                amountCents,
                currency
            );
        }

        private Map<String, Object> upcastOnce(
            String eventType,
            int fromVersion,
            Map<String, Object> payload
        ) {
            if (fromVersion == 1 && ("FundsDeposited".equals(eventType) || "FundsWithdrawn".equals(eventType))) {
                Map<String, Object> upgraded = new LinkedHashMap<>();
                upgraded.put("amount_cents", ((Number) payload.get("amount_cents")).intValue());
                upgraded.put("currency", "USD");
                return upgraded;
            }
            throw new IllegalArgumentException("missing upcaster for " + eventType + " v" + fromVersion);
        }
    }

    public static final class AccountProjector {
        private int balanceCents = 0;

        public void apply(CanonicalEvent event) {
            if (!"USD".equals(event.getCurrency())) {
                throw new IllegalArgumentException("unsupported currency: " + event.getCurrency());
            }
            switch (event.getEventType()) {
                case "FundsDeposited":
                    balanceCents += event.getAmountCents();
                    return;
                case "FundsWithdrawn":
                    balanceCents -= event.getAmountCents();
                    return;
                default:
                    throw new IllegalArgumentException("unsupported eventType: " + event.getEventType());
            }
        }

        public int getBalanceCents() {
            return balanceCents;
        }
    }

    public static final class CommandService {
        private static final int LATEST_VERSION = 2;
        private final EventStore store;

        public CommandService(EventStore store) {
            this.store = store;
        }

        public void deposit(String accountId, int amountCents, String currency) {
            if (amountCents <= 0) {
                throw new IllegalArgumentException("amountCents must be positive");
            }
            Map<String, Object> payload = new LinkedHashMap<>();
            payload.put("amount_cents", amountCents);
            payload.put("currency", currency);
            store.append(accountId, "FundsDeposited", LATEST_VERSION, payload);
        }

        public void withdraw(String accountId, int amountCents, String currency) {
            if (amountCents <= 0) {
                throw new IllegalArgumentException("amountCents must be positive");
            }
            Map<String, Object> payload = new LinkedHashMap<>();
            payload.put("amount_cents", amountCents);
            payload.put("currency", currency);
            store.append(accountId, "FundsWithdrawn", LATEST_VERSION, payload);
        }
    }

    public static int rebuildBalance(
        String accountId,
        EventStore store,
        EventUpcaster upcaster
    ) {
        AccountProjector projector = new AccountProjector();
        for (RawEvent rawEvent : store.eventsForAccount(accountId)) {
            CanonicalEvent canonical = upcaster.upcast(rawEvent);
            projector.apply(canonical);
        }
        return projector.getBalanceCents();
    }
}

File: Day-22/event-upcasting-sample/python/event_upcasting_ledger.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class RawEvent:
    sequence: int
    account_id: str
    event_type: str
    schema_version: int
    payload: dict[str, Any]


@dataclass(frozen=True)
class CanonicalEvent:
    sequence: int
    account_id: str
    event_type: str
    schema_version: int
    amount_cents: int
    currency: str


class EventStore:
    def __init__(self) -> None:
        self._events: list[RawEvent] = []
        self._next_sequence = 1

    def append(
        self,
        account_id: str,
        event_type: str,
        schema_version: int,
        payload: dict[str, Any],
    ) -> RawEvent:
        if not account_id:
            raise ValueError("account_id must be non-empty")
        if not event_type:
            raise ValueError("event_type must be non-empty")
        if schema_version <= 0:
            raise ValueError("schema_version must be positive")

        event = RawEvent(
            sequence=self._next_sequence,
            account_id=account_id,
            event_type=event_type,
            schema_version=schema_version,
            payload=payload,
        )
        self._events.append(event)
        self._next_sequence += 1
        return event

    def events_for_account(self, account_id: str) -> list[RawEvent]:
        return [event for event in self._events if event.account_id == account_id]


class EventUpcaster:
    _LATEST_VERSION_BY_TYPE = {
        "FundsDeposited": 2,
        "FundsWithdrawn": 2,
    }

    def upcast(self, event: RawEvent) -> CanonicalEvent:
        latest = self._LATEST_VERSION_BY_TYPE.get(event.event_type)
        if latest is None:
            raise ValueError(f"unsupported event_type: {event.event_type}")
        if event.schema_version > latest:
            raise ValueError(
                f"unsupported future schema for {event.event_type}: {event.schema_version} > {latest}"
            )

        current_version = event.schema_version
        payload = dict(event.payload)

        while current_version < latest:
            payload = self._upcast_once(event.event_type, current_version, payload)
            current_version += 1

        amount_cents = int(payload["amount_cents"])
        currency = str(payload["currency"])
        return CanonicalEvent(
            sequence=event.sequence,
            account_id=event.account_id,
            event_type=event.event_type,
            schema_version=current_version,
            amount_cents=amount_cents,
            currency=currency,
        )

    @staticmethod
    def _upcast_once(
        event_type: str, from_version: int, payload: dict[str, Any]
    ) -> dict[str, Any]:
        if from_version == 1 and event_type in {"FundsDeposited", "FundsWithdrawn"}:
            return {
                "amount_cents": int(payload["amount_cents"]),
                "currency": "USD",
            }
        raise ValueError(f"missing upcaster for {event_type} v{from_version}")


class AccountProjector:
    def __init__(self) -> None:
        self.balance_cents = 0

    def apply(self, event: CanonicalEvent) -> None:
        if event.currency != "USD":
            raise ValueError(f"unsupported currency: {event.currency}")

        if event.event_type == "FundsDeposited":
            self.balance_cents += event.amount_cents
            return
        if event.event_type == "FundsWithdrawn":
            self.balance_cents -= event.amount_cents
            return
        raise ValueError(f"unsupported event_type: {event.event_type}")


class CommandService:
    _LATEST_VERSION = 2

    def __init__(self, event_store: EventStore) -> None:
        self._event_store = event_store

    def deposit(self, account_id: str, amount_cents: int, currency: str = "USD") -> None:
        if amount_cents <= 0:
            raise ValueError("amount_cents must be positive")
        self._event_store.append(
            account_id=account_id,
            event_type="FundsDeposited",
            schema_version=self._LATEST_VERSION,
            payload={"amount_cents": amount_cents, "currency": currency},
        )

    def withdraw(self, account_id: str, amount_cents: int, currency: str = "USD") -> None:
        if amount_cents <= 0:
            raise ValueError("amount_cents must be positive")
        self._event_store.append(
            account_id=account_id,
            event_type="FundsWithdrawn",
            schema_version=self._LATEST_VERSION,
            payload={"amount_cents": amount_cents, "currency": currency},
        )


def rebuild_balance(account_id: str, store: EventStore, upcaster: EventUpcaster) -> int:
    projector = AccountProjector()
    for raw_event in store.events_for_account(account_id):
        canonical = upcaster.upcast(raw_event)
        projector.apply(canonical)
    return projector.balance_cents
